csharp/src/Google.Protobuf/CodedInputStream.cs

```diff
@@ -470,87 +521,259 @@ internal void SkipGroup(uint startGroupTag)
         /// <summary>
         /// Reads a double field from the stream.
         /// </summary>
+        [SecuritySafeCritical]
         public double ReadDouble()
         {
-            return BitConverter.Int64BitsToDouble((long) ReadRawLittleEndian64());
+            var immediateBudder = ImmediateBuffer;
+            return ReadDouble(ref immediateBudder);
         }
 
         /// <summary>
-        /// Reads a float field from the stream.
+        /// This supports the Protocol Buffers infrastructure and is not meant to be used directly from your code.
         /// </summary>
-        public float ReadFloat()
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [SecurityCritical]
+        public double ReadDouble(ref ReadOnlySpan<byte> immediateBuffer) => BitConverter.Int64BitsToDouble((long)ReadRawLittleEndian64(ref immediateBuffer));
+
+        /// <summary>
+        /// This supports the Protocol Buffers infrastructure and is not meant to be used directly from your code.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [SecurityCritical]
+        public double ReadWrappedDouble(ref ReadOnlySpan<byte> immediateBuffer)
         {
-            if (BitConverter.IsLittleEndian && 4 <= bufferSize - bufferPos)
+            int oldLimit = BeginReadNested(ref immediateBuffer);
+            uint tag;
+            double value = default(double);
+            while ((tag = ReadTag(ref immediateBuffer)) != 0)
             {
-                float ret = BitConverter.ToSingle(buffer, bufferPos);
-                bufferPos += 4;
-                return ret;
+                if (tag == WellKnownTypes.WrappersReflection.WrapperValueFixed64Tag)
+                {
+                    value = ReadDouble(ref immediateBuffer);
+                }
+                else
+                {
+                    SkipLastField(ref immediateBuffer);
+                }
             }
-            else
+            EndReadNested(oldLimit);
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a float field from the stream.
+        /// </summary>
+        [SecuritySafeCritical]
+        public float ReadFloat()
+        {
+            var immediateBudder = ImmediateBuffer;
+            return ReadFloat(ref immediateBudder);
+        }
+
+        /// <summary>
+        /// This supports the Protocol Buffers infrastructure and is not meant to be used directly from your code.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [SecurityCritical]
+#if NETCOREAPP2_1
+        public float ReadFloat(ref ReadOnlySpan<byte> immediateBuffer) => BitConverter.Int32BitsToSingle((int)ReadRawLittleEndian32(ref immediateBuffer));

```

Comment:
readability: I think it's better to use {   } code block and use #if  #else inside that code block.  That also guarantees that the method signature will be that same for all framework.